cmds/id/id.go

```diff
@@ -0,0 +1,265 @@
+		return fmt.Errorf("cannot print \"only\" of more than one choice\n")
+
+	}
+	if flags.n && !(flags.G || flags.g || flags.u) {
+		return fmt.Errorf("cannot print only names in default format\n")
+	}
+
+	return nil
+}
+
+type User struct {
+	Name   string
+	Uid    int
+	Euid   int
+	Gid    int
+	Groups map[int]string
+}
+
+func (u *User) GetUid() int {
+	return u.Uid
+}
+
+func (u *User) pullUid() {
+	u.Uid = syscall.Getuid()
+}
+
+func (u *User) GetEuid() int {
+	return u.Euid
+}
+
+func (u *User) pullEuid() {
+	u.Uid = syscall.Geteuid()
+}
+
+func (u *User) GetGid() int {
+	return u.Gid
+}
+
+func (u *User) pullGid() {
+	u.Gid = syscall.Getgid()
+}
+
+func (u *User) GetName() string {
+	return u.Name
+}
+
+// pullName finds the name of the User by reading PASSWD_FILE.
+func (u *User) pullName() error {
+	passwdFile, err := os.Open(PASSWD_FILE)
+	if err != nil {
+		return err
+	}
+
+	var passwdInfo []string
+
+	passwdScanner := bufio.NewScanner(passwdFile)
+
+	for passwdScanner.Scan() {
+		passwdInfo = strings.Split(passwdScanner.Text(), ":")
+		if val, err := strconv.Atoi(passwdInfo[2]); err != nil {
+			return err
+		} else if val == u.GetUid() {
+			u.Name = passwdInfo[0]
+			return nil
+		}
+	}
+
+	return fmt.Errorf("User is not in %s", PASSWD_FILE)
+}
+
+func (u *User) GetGroups() map[int]string {
+	return u.Groups
+}
+
+func (u *User) GetGidName() string {
+	val := u.GetGroups()[u.GetUid()]
+	return val
+}
+
+// pullGroups aggregates the groups that the User is in.
+func (u *User) pullGroups() error {
+	groupsNumbers, err := syscall.Getgroups()
+	if err != nil {
+		return err
+	}
+
+	groupsMap, err := readGroups()
+	if err != nil {
+		return err
+	}
+
+	for _, groupNum := range groupsNumbers {
+		if groupName, ok := groupsMap[groupNum]; ok {
+			u.Groups[groupNum] = groupName
+		} else {
+			return fmt.Errorf("Inconsistent %s file", GROUP_FILE)
+		}
+	}
+	return nil
+}
+
+// readGroups reads the GROUP_FILE for groups.
+// It assumes the format "name:passwd:number:groupList".
+func readGroups() (map[int]string, error) {
+	groupFile, err := os.Open(GROUP_FILE)
+	if err != nil {
+		return nil, err
+	}
+
+	var groupInfo []string
+
+	groupsMap := make(map[int]string)
+	groupScanner := bufio.NewScanner(groupFile)
+
+	for groupScanner.Scan() {
+		groupInfo = strings.Split(groupScanner.Text(), ":")
+		groupNum, err := strconv.Atoi(groupInfo[2])
+		if err != nil {
+			return nil, err
+		}
+
+		groupsMap[groupNum] = groupInfo[0]
+	}
+
+	return groupsMap, nil
+}
+
+// NewUser is a factory method for the User type.
+func NewUser() (*User, error) {
+	emptyMap := make(map[int]string)
+	u := &User{"", -1, -1, -1, emptyMap}
+	u.pullUid()
+	u.pullEuid()
+	u.pullGid()
+	if err := u.pullGroups(); err != nil {
+		return nil, err
+	}
+	if err := u.pullName(); err != nil {
+		return nil, err
+	}
+	return u, nil
+}
+
+// IDCommand runs the "id" with the current user's information.
+func IDCommand(u User) {
+	if !flags.G {
+		if flags.u {
+			if flags.n {
```

Comment:
I have no idea what these flags mean, tbh. +1 to the descriptive names. "if g and u and n" is not comprehensible to me